refactor: move contacts out of the wallet

brianp opened the pull request:
Description
---
Refactor contacts and the contact service out of the wallet crate and into it's own crate to be included anywhere needed.

Motivation and Context
---
Wallet doesn't need to manage contacts, and we'll want contacts for other applications so having it as it's own crate makes it portable.

How Has This Been Tested?
---
~Entirely untested. I haven't run this code, only trusting the compiler. Do not merge.~

Tested manually by running the wallet. Many failing tests caught and corrected via CI.

brianp committed:
Extract Contacts and Contact Service

Move contacts out of the wallet and into its own crate. This allows
anyone the ability to utilize the contacts crate for their own
applications. Move useful sqlite related types to the common crate.

stringhandler commented on base_layer/contacts/Cargo.toml:
```diff
@@ -0,0 +1,73 @@
+authors = ["The Tari Development Community"]
+description = "Tari contacts library"
+license = "BSD-3-Clause"
+version = "0.47.0-pre.0"
```
Please add this file to `scripts/update_crate_metadata.sh`

stringhandler reviewed (commented):
Looking good so far

brianp committed:
Pruning Cargo.toml

brianp committed:
Make sure the new crate will have its version updated

brianp commented on base_layer/contacts/src/lib.rs:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2020, The Tari Project
+// Copyright 2023. The Tari Project
```
```suggestion
// Copyright 2023, The Tari Project
```

brianp commented on base_layer/wallet_ffi/wallet.h:
```diff
@@ -68,10 +68,6 @@ struct CommitmentAndPublicKeySignature_RistrettoPublicKey__RistrettoSecretKey;
 
 struct CompletedTransaction;
 
-struct Contact;
-
-struct ContactsLivenessData;
-
```
I feel like this is a failed merge/conflict.

brianp replied:
hmmm this happens on compile, so I guess not a conflict.

brianp commented on base_layer/wallet_ffi/wallet.h:
```diff
@@ -302,9 +298,9 @@ typedef struct Covenant TariCovenant;
+typedef Contact TariContact;
 
-typedef struct ContactsLivenessData TariContactsLivenessData;
+typedef ContactsLivenessData TariContactsLivenessData;
```
Same. This looks like a conflict.

brianp committed:
Use a generic DbConnection where possible

Have the sqlite common crate provide a generic db connection struct for
use in simplified cases such as testing, and hopefully we could move all
or most connection types over to this in the future.

Moves the error type of PooledDbConnection from the contacts crate over
to sqlite_common as it's more appropriate in the common crate.

brianp committed:
Update already moved deps

brianp committed:
Remove wallet dependecies from contacts tests

brianp committed:
Change ContactsService initialization for wallet and ffi wallet

brianp commented:
@stringhandler this is _finally_ ready for a real review.

sdbondi reviewed (approved):
utACK

Nice and neat :)

SWvheerden reviewed (approved):
utACK

SWvheerden commented:
One thing we might want to decide, for the next PR. 
Is what do we do about the address? Currently, its all tied to the wallet via a `TariAddress`. This means the contacts are all tied per network. Perhaps this is what we want, or perhaps we need to change this to `public key` or both?

stringhandler commented on common_sqlite/src/error.rs:
```diff
@@ -22,12 +22,45 @@
+}
+
+#[derive(Debug, Error)]
+pub enum StorageError {
```
At some point we can merge this and the above. The diesel errors are too general to have as "StorageError"
